### ISSUE TITLE
Proposal: document structure of annotation/label namespaces more completely

### ISSUE THREAD

**Issue by thockin** [kubernetes/community#300]
Since we have ever more incubated projects, and everyone loves annotations and labels, we need a spec for how those annotations are formed.

We also have extant names, such as volume.alpha.kubernetes.io, pod.alpha.kubernetes.io, beta.kubernetes.io, net.beta.kubernetes.io, as well as deployment.kubernetes.io, kubectl.kubernetes.io, and vendor or product-centric quobyte.kubernetes.io, rkt.kubernetes.io, and gluster.kubernetes.io that we need to cover.  

Proposal:

* `kubernetes.io/`, and `*.kubernetes.io` (collectively called
top-level names) are managed centrally and should be considered
reserved.  We will document and manage them centrally.

* `*.alpha.kubernetes.io`, and `*.beta.kubernetes.io` are reserved for
alpha/beta of centrally managed names.

* `*.<repo>.x.kubernetes.io` is reserved for use in that component /
repo (NB, repo moves will cause pain).

* As a component becomes more popular, it may petition for a top-level
(non-x) name.

* All existing names will be grandfathered and not forced to change, though owners are encouraged to accept and document the "new" names.

@philips lips @justinsb @bgrant0607 @smarterclayton @brendandburns

**Comment by justinsb** [comment]
I like this suggestion.  I have tried a few times, the biggest problem I've hit seems to be that we can't agree where to put things.  If labels are docs, they get bounced around outside the main repo and ignored.  I think they are code, but we need a clear dictat that this will happen and that the existing labels shall be added as-is - i.e. not a time to re-litigate existing labels.

For repos, I think it's important to think carefully about the name before proposing it for incubation.  We should have some guidelines to avoid obvious pitfalls we've hit (e.g. no "cute" names, no codenames, run it by some politically-correct people).

What I don't understand is the `.x.` bit:  I don't see what it really gains us, other than reflecting our organizational structure, and in particular our github structure.  That seems wrong - if we switch to gerrit and discover that a monorepo is workable, we wouldn't rename our labels, for example.   The cost is clear: at some stage we expect users to change labels because a project has become popular, and we do this just when a project is widely used, which seems inopportune timing.

I'd suggest instead we drop the 'x' entirely, and say that some `*.kubernetes.io` names are administered in the kubernetes/kubernetes repo (today).  But the repo name is not tied to the label name, though obviously having them be close minimizes confusion, and I think an incubated project _should_ get a name (e.g. `external-dns.kubernetes.io`), to avoid conflicts.

In practice, I think this means having a doc in kubernetes/kubernetes which lists the partitions and has links to a doc (in the same repo or a separate repo) where the labels are managed.  DNS in markdown.

An obvious downside is that if an incubation project does not graduate, then that name is forever removed from the namepool.  But the same problem applies even with `x`, and also I don't think we would reuse labels - it is at best confusing, and if old labels suddenly started doing things again that would be surprising.

**Comment by philips** [comment]
Overall I think this is a good proposal. My desired outcomes for creating a well-known label registry:

- Increase user and community discoverability of labels
- Have consistency around deprecation (2 versions, 1 year, whatever)
- Documentation on best practices
- Encourage good patterns through review
- Provide a Go package for import

For projects I agree that .x. is sort of useless; just like the old [X- header in HTTP](https://tools.ietf.org/html/rfc6648#section-3). An alternative idea would be that projects simply register a prefix and their git repo; and if they use the name of their project no approval is necessary:

- github.com/kubernetes-incubator/nfs-provisioner gets incubator-nfs-provisioner.kubernetes.io and github.com/kubernetes/kops gets kops.kubernetes.io.
- Any {alpha,beta}.kops.kubernetes.io and {alpha,beta}.incubator-nfs-provisioner.kubernetes.io SHOULD be registered but are not required to be
- Any incubator-nfs-provisioner.kubernetes.io or kops.kubernetes.io labels MUST be registered

**Comment by thockin** [comment]
On Fri, Jan 27, 2017 at 1:54 PM, Brandon Philips
<notifications@github.com> wrote:
>
> Overall I think this is a good proposal. My desired outcomes for creating a well-known label registry:
>
> Increase user and community discoverability of labels
> Have consistency around deprecation (2 versions, 1 year, whatever)
> Documentation on best practices
> Encourage good patterns through review
> Provide a Go package for import
>
> For projects I agree that .x. is sort of useless; just like the old X- header in HTTP. An alternative idea would be that projects simply register a prefix and their git repo; and if they use the name of their project no approval is necessary:

The .x. was specifically to normalize the existing names
(uncoordinated) with the complete namespace of every repo possible.
It's unlikely, but not impossible that a repo called "scheduler" could
be created and then that is a conflict.  The .x. specifically says
this is not "mainstream" yet, which gives us more freedom to make
mistakes.

I'm not 100% sold on it, but it seemed safe - ESPECIALLY for incubator projects.

> github.com/kubernetes-incubator/nfs-provisioner gets incubator-nfs-provisioner.kubernetes.io

See, you're namespacing too.  Where I said ".x." you said
"incubator-".  Properly, your proposal should be
`nfs-provisioner.incubator.kubernetes.io`, since incubation is a
higher-order datum.  At that point, I argue .x. is simply shorter.

> and github.com/kubernetes/kops gets kops.kubernetes.io.

I am probably OK with repo names in the main org being considered the
centrally-managed name.

As to where it lives, I think living in code is good.  Main repo or a
while new repo doesn't matter much.  Either way it needs humans to
police it.  If we get used to the idea that literal values only appear
in docs, and all code uses const symbols, it could be winnable.

**Comment by smarterclayton** [comment]
I'm on board with this rough structure.  Where disputes in naming for `.x.` arise (prior or post incubation), we should probably set the bar that:

* Would a developer on the project associate the name with a sig AND
* Did that SIG get a chance to give feedback on naming.

The incubator process is effectively a DNS registrar for top level names, so even today it's a good point of control.

**Comment by philips** [comment]
On Sat, Jan 28, 2017 at 8:55 AM Tim Hockin <notifications@github.com> wrote:

> I'm not 100% sold on it, but it seemed safe - ESPECIALLY for incubator
> projects.
>

OK, +1 on .x. Forget my incubator- idea.


> > and github.com/kubernetes/kops gets kops.kubernetes.io.
>
> As to where it lives, I think living in code is good. Main repo or a
> while new repo doesn't matter much. Either way it needs humans to
> police it. If we get used to the idea that literal values only appear
> in docs, and all code uses const symbols, it could be winnable.


The rough thing I think will work:

1) Registered label package: People can register a new label namespace
(e.g. foobar.kubernetes.io) an email to a SIG with a few interested parties
and then get a PR approved by a label registry maintainer. The label
maintainers help everyone avoid bad label patterns and ensure important top
level names don't get claimed (e.g. scheduler).

2) API Well Known: If a label starts to get wide use over 6-12 months we
can put it into the API as a well known.

Thoughts on that?

**Comment by justinsb** [comment]
I don't want to hijack this proposal, but should we discuss alpha & beta in labels & annotations here?

For example, on https://github.com/kubernetes/kubernetes/pull/40088 , we end up with this:

> Together with the API endpoint, StorageClass annotation "storageclass.beta.kubernetes.io/is-default-class" is deprecated and  "storageclass.kubernetes.io/is-default-class" should be used instead to mark a default storage class. The beta annotation is still working in this release, however it won't be supported in the next one.

(Nothing particularly criticizing that PR, that is just the natural consequence).

That means we effectively break user manifest files.  We do all this API compatibility work to make sure everything works across versions, but then it breaks because of an annotation.

Is there something we can do to make this easier on users?  Some ideas:

* Don't include `alpha` or `beta` in the annotation names
* Upgrade the annotation names automatically (or auto populate fields)
* Recognize alpha / beta annotation names "forever" / for a long time
* Provide some linter that does static analysis (maybe part of kubectl)
* Provide some runtime advisor that logs events or similar

**Comment by luxas** [comment]
Also see: https://github.com/kubernetes/kubernetes/pull/41835

**Comment by deads2k** [comment]
I think that we face a similar problem with namespace names.  To prevent future pain, we could reserve all namespaces that start with `kube-`.

We currently create a couple well known namespaces called `kube-system` and `kube-public`.  I suspect that over time we may develop more, particularly as we start self-hosting pieces of infrastructure or want to extend core functionality with "blessed" add-ons.  Claiming the `kube-*` pattern for kube now will alleviate future problems.

@kubernetes/sig-api-machinery-misc @smarterclayton @liggitt @lavalamp

**Comment by luxas** [comment]
@deads2k Agree. But how do you envision we can claim those namespaces?
Do you mean we should restrict the user from creating other `kube-*` namespaces or do you just mean that we should document them as claimed/well-known?

**Comment by deads2k** [comment]
> @deads2k Agree. But how do you envision we can claim those namespaces?
> Do you mean we should restrict the user from creating other kube-* namespaces or do you just mean that we should document them as claimed/well-known?

Document the intent. I don't think we can reasonably or easily actively restrict their creation with our current infrastructure.

**Comment by luxas** [comment]
Yeah, that's what I thought. Seems like https://kubernetes.io/docs/api-reference/labels-annotations-taints/ or the like is the right place to start...

**Comment by mrbobbytables** [comment]
Is this still relevant? It's been open and frozen for some time now.
I'm going to mark it as rotten for now, if it should be left open please update or re-freeze it 👍  

/remove-lifecycle frozen
/lifecycle rotten

**Comment by lavalamp** [comment]
@deads2k mostly fixed this for group names, IMO.

But labels still have this problem?

**Comment by thockin** [comment]
This is still relevant.  It's something we actually need to do.
